src/tutorial/function-dispatching/README.md

```diff
@@ -0,0 +1,352 @@
+
+This method seems naive, however this is exactly how Vyper and Solidity\* implement linear dispatching. If you want it to be cheaper to call, just move it higher up in the contract!
+
+\* Solidity only implements this method if there are less than 4 functions in a contract.
+
+You may be wondering what the series of jump labels followed by implementation macros mean. A good mental model is that the entire of your contract exists within your MAIN macro. If a macro is not referred to within main or nested within a macro that is invoked within MAIN then it will not be included in your contract. In reality the above contract when compiled looks like this:
+
+```
+0x[dispatcher]<allowanceJumpLabel>[ALLOWANCE MACRO]<approveJumpLabel>[APPROVE MACRO]<balanceOfJump>[BALANCE_OF MACRO] ... <transferFromLabel>[TRANSFER_FROM MACRO]<transferJump>[TRANSFER MACRO]
+```
+
+In Huff all of your macros are inlined into one long bytecode string (this is also true for Vyper and Solidity!). Due to this property it is salient that you terminate ALL of your top-level macros with some form of escape opcode (`return`, `stop`, `revert`). If you don't do this your macro _WILL_ continue to execute which ever macro is inlined next until a return condition is found.
+
+For example:
+
+```huff
+#define function returnOne() public returns(uint256)
+#define function returnTwo() public returns(uint256)
+
+/// @notice returns the value one
+#define macro RETURN_ONE() = {
+    0x01 0x00 mstore
+    0x20 0x00 return
+}
+
+/// @notice places the value 2 onto the stack
+#define macro RETURN_TWO() = {
+    0x02
+}
+
+#define macro MAIN() = {
+    0x00 calldataload 0xE0 shr
+
+    // Dispatcher
+    dup1 __FUNC_SIG(returnOne) eq returnOneJump jumpi
+    dup1 __FUNC_SIG(returnTwo) eq returnTwoJump jumpi
+
+    // Macros
+    returnTwo:
+        RETURN_TWO()
+    returnOne:
+        RETURN_ONE()
+}
+```
+
+In the above macro, regardless if what message is sent to it, it will ALWAYS return the value 1. As the macro `RETURN_TWO` does not terminate, it will roll over into the execution of `RETURN_ONE`, terminating with returning the value one. As there is no guard to protect against non valid function signatures any message, even if it matches nothing in the dispatcher will return the value 1.
+
+To protect against no valid function selector being found we can include `0x00 0x00 revert` in-between our dispatcher and our macro jump labels as so:
+
+```huff
+#define macro MAIN() = {
+    0x00 calldataload 0xE0 shr
+
+    dup1 __FUNC_SIG(returnOne) eq returnOneJump jumpi
+    dup1 __FUNC_SIG(returnTwo) eq returnTwoJump jumpi
+
+    0x00 0x00 revert
+
+    returnTwo:
+        RETURN_TWO()
+    returnOne:
+        RETURN_ONE()
+}
+```
+
+Please keep this behavior in mind when writing Huff contracts, especially when dealing with administration functions, running into a `SET_OWNER()` macro could allow anyone to take control of your contract.
```

Comment:
We can just add that new section on inheritance later and leave this vague reference for now.
```suggestion
Please keep this behavior in mind when writing Huff contracts, especially when dealing with administration functions, running into a `SET_OWNER()` macro could allow anyone to take control of your contract.  There is a more advanced way of jumping over the dispatcher that is used with Huff "inheritance" pattern that we will discuss in another section.
```